### PR TITLE
feat: css parser

### PR DESCRIPTION
**Pull request by bitpshr**
## Summary

This pull request adds a tolerant CSS parser for the base [CSS3 specification](https://www.w3.org/TR/css-syntax-3/#tokenization). 

CSS3 is made up of various sub-grammars for things like selectors and individual at-rules. These grammars specify rule-specific parsing algorithms that cannot be uniformly-defined in the base CSS specification itself, and they can be very complex and may warrant one-off parsers. This base parser implementation only parses nodes per the base [CSS3 specification](https://www.w3.org/TR/css-syntax-3/#tokenization) and does not include per-rule grammar support required for "full" CSS3 support. These rules don't throw errors and are still tokenized and parsed, but grammar-specific nodes and error-checking are not yet supported. My thinking was that support for these sub-grammars can be added iteratively going forward.

### Remaining tasks for full CSS support

1. Implement sub-grammar parsers. I can work to identify these and to ticket them out individually so anyone can pick them up.
1. Add builders for new types.

## Test Plan

- Verify the output in `packages/@romefrontend/css-parser/tests-fixtures/smoke/input.test.md`

### CODE REVIEW

File: packages/@romefrontend/css-parser/types.ts [hunk 1]
```diff
@@ -35,8 +44,10 @@ export type Tokens = BaseTokens & {
 	BadString: SimpleToken<"BadString">;
 	BadURL: SimpleToken<"BadURL">;
 	CDC: SimpleToken<"CDC">;
+	CDO: SimpleToken<"CDO">;
```
Comment (ematipico) [inline packages/@romefrontend/css-parser/types.ts:47]: Out of curiosity: what CDO and CDC stand for?

**Comment by ematipico** [comment]
I am aware that this is not finished and the parser will grow a lot! And the other parsers too! I was wondering if we should split up the code somehow? Or is it fine to have everything in one single file?

**Comment by bitpshr** [comment]
Thanks for the review @ematipico. I agree that we should better modularize all parsers; I think most are in huge files like this one.

@sebmck, any thoughts on this approach and the described path forward for CSS support? The complexity of the various grammars that make up different aspects of CSS3 was admittedly frustrating. Still, we could technically begin to provide CSS formatting and linting, it would just be more manual and less robust in terms of AST <-> spec parity. For example, selector-based lint rules could still be added since we're at least tokenizing selector parts. The same is true for rules for keyframes and other at-rules. Fully tokenizing these nodes into the highest-level tokens defined by each sub-grammar (and verifying them as such) isn't necessarily a prerequisite for beginning to lint and format CSS.

File: packages/@romefrontend/ast-utils/isRoot.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ export function isRoot(node: TransformExitResult): node is AnyRoot {
 		!Array.isArray(node) &&
 		typeof node !== "symbol" &&
 		(node.type === "JSRoot" ||
-		node.type === "CSSRoot" ||
+		node.type === "CSSStylesheet" ||
```
Comment (sebmck) [inline packages/@romefrontend/ast-utils/isRoot.ts:9]: We should still call the node `CSSRoot` here to maintain consistency.

File: packages/@romefrontend/ast/css/core/CSSAtRule.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import {CSSBlock, NodeBaseWithComments} from "@romefrontend/ast";
+import {createBuilder} from "@romefrontend/ast/utils";
+import {AnyCSSValue} from "../../../css-parser/types";
+
+export type CSSAtRule = NodeBaseWithComments & {
+	type: "CSSAtRule";
+	name: string;
+	prelude: Array<AnyCSSValue>;
+	block: CSSBlock | null;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSAtRule.ts:9]: We actually don't use `null` anywhere in Rome, except when checking values given to use by Node or native methods.

```suggestion
	block?: CSSBlock;
```

File: packages/@romefrontend/ast/css/core/CSSBlock.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	CSSDeclaration,
+	CSSRule,
+	NodeBaseWithComments,
+} from "@romefrontend/ast";
+import {createBuilder} from "@romefrontend/ast/utils";
+import {AnyCSSValue} from "../../../css-parser/types";
+
+export type CSSBlock = NodeBaseWithComments & {
+	type: "CSSBlock";
+	startingTokenValue: string | null;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSBlock.ts:12]:
```suggestion
	startingTokenValue?: string;
```

File: packages/@romefrontend/ast/css/core/CSSBlock.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	value?: Array<AnyCSSValue | CSSRule | CSSAtRule | CSSDeclaration | null>;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSBlock.ts:13]:
```suggestion
	value?: Array<AnyCSSValue | CSSRule | CSSAtRule | CSSDeclaration | undefined>;
```

File: packages/@romefrontend/ast/css/core/CSSDeclaration.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import {NodeBaseWithComments} from "@romefrontend/ast";
+import {createBuilder} from "@romefrontend/ast/utils";
+import {AnyCSSValue} from "../../../css-parser/types";
+
+export type CSSDeclaration = NodeBaseWithComments & {
+	type: "CSSDeclaration";
+	name: string;
+	important: boolean;
+	value: Array<AnyCSSValue | null>;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSDeclaration.ts:9]:
```suggestion
	value: Array<AnyCSSValue | undefined>;
```

File: packages/@romefrontend/ast/css/core/CSSFunction.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import {NodeBaseWithComments} from "@romefrontend/ast";
+import {createBuilder} from "@romefrontend/ast/utils";
+import {AnyCSSValue} from "../../../css-parser/types";
+
+export type CSSFunction = NodeBaseWithComments & {
+	type: "CSSFunction";
+	name: string;
+	value: Array<AnyCSSValue | null>;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSFunction.ts:8]:
```suggestion
	value: Array<AnyCSSValue | undefined>;
```

File: packages/@romefrontend/ast/css/core/CSSRule.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import {CSSBlock, NodeBaseWithComments} from "@romefrontend/ast";
+import {createBuilder} from "@romefrontend/ast/utils";
+import {AnyCSSValue} from "../../../css-parser/types";
+
+export type CSSRule = NodeBaseWithComments & {
+	type: "CSSRule";
+	prelude: Array<AnyCSSValue>;
+	block: CSSBlock | null;
```
Comment (sebmck) [inline packages/@romefrontend/ast/css/core/CSSRule.ts:8]:
```suggestion
	block?: CSSBlock;
```

File: packages/@romefrontend/css-parser/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+	const outputFile = inputFile.absolute.getParent().append(
+		inputFile.absolute.getExtensionlessBasename(),
+	).join();
+
+	t.snapshot(ast, undefined, {filename: outputFile});
+});
+
+// @ts-ignore Doesn't support top-level await
+await promise;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/index.test.ts:30]: As long as this is declared as a file handler with a `parse` method you can replace this with:

```suggestion
import {declareParserTests} from "@romefrontend/test-helpers";

const promise = declareParserTests();

// @ts-ignore Doesn't support top-level await lol
await promise;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 10]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+			});
+			return null;
+		}
+
+		parseAtRule(): CSSAtRule {
+			const start = this.getPosition();
+			const token = this.expectToken("AtKeyword");
+			const prelude: Array<AnyCSSValue> = [];
+			const name = token.value;
+			let block = null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:701]:
```suggestion
			let block;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 11]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+					continue;
+				}
+				const rule = this.parseRule();
+				rule && rules.push(rule);
+			}
+
+			return rules;
+		}
+
+		parseRule(): CSSRule | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:671]:
```suggestion
		parseRule(): CSSRule | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 12]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+						},
+					);
+				}
+				const parsedValue = this.parseComponentValue();
+				parsedValue && prelude.push(parsedValue);
+			}
+			this.unexpectedDiagnostic({
+				description: descriptions.CSS_PARSER.UNEXPECTED_TOKEN,
+			});
+			return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:693]:
```suggestion
			return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 13]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				{
+					type: "CSSAtRule",
+					name,
+					prelude,
+					block,
+				},
+			);
+		}
+
+		parseSimpleBlock(): CSSBlock | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:730]:
```suggestion
		parseSimpleBlock(): CSSBlock | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 14]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+			const start = this.getPosition();
+			const startingToken = this.getToken();
+			const startingTokenValue = this.getBlockStartTokenValue(startingToken);
+			const endingTokenType = this.getBlockEndTokenType(startingToken);
+			let value: Array<AnyCSSValue | CSSAtRule | CSSDeclaration | null> = [];
+
+			if (!endingTokenType) {
+				return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:738]:
```suggestion
				return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 15]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				start,
+				{
+					type: "CSSBlock",
+					startingTokenValue,
+					value,
+				},
+			);
+		}
+
+		parseDeclartionBlock(): CSSBlock | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:768]:
```suggestion
		parseDeclartionBlock(): CSSBlock | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 16]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+			const start = this.getPosition();
+			const startingToken = this.getToken();
+			const startingTokenValue = this.getBlockStartTokenValue(startingToken);
+			const endingTokenType = this.getBlockEndTokenType(startingToken);
+			let value: Array<AnyCSSValue | CSSAtRule | CSSDeclaration | null> = [];
+
+			if (!endingTokenType) {
+				return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:776]:
```suggestion
				return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 17]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				start,
+				{
+					type: "CSSBlock",
+					startingTokenValue,
+					value,
+				},
+			);
+		}
+
+		parseComplexBlock(): CSSBlock | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:793]:
```suggestion
		parseComplexBlock(): CSSBlock | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 18]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+			const start = this.getPosition();
+			const startingToken = this.getToken();
+			const startingTokenValue = this.getBlockStartTokenValue(startingToken);
+			const endingTokenType = this.getBlockEndTokenType(startingToken);
+			let value: Array<CSSAtRule | CSSRule> = [];
+
+			if (!endingTokenType) {
+				return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:801]:
```suggestion
				return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 19]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+					startingTokenValue,
+					value,
+				},
+			);
+		}
+
+		parseComponentValue(): AnyCSSValue | null {
+			if (this.matchToken("Whitespace")) {
+				this.nextToken();
+				return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:821]:
```suggestion
				return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 20]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				});
+				while (!this.matchToken("Semi") && !this.matchToken("EOF")) {
+					this.parseComponentValue();
+				}
+			}
+
+			return declarations;
+		}
+
+		parseDeclaration(): CSSDeclaration | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:971]:
```suggestion
		parseDeclaration(): CSSDeclaration | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 21]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+			while (!this.matchToken("Semi")) {
+				const currentToken = this.getToken();
+				if (currentToken.type !== "Ident") {
+					this.unexpectedDiagnostic({
+						description: descriptions.CSS_PARSER.INVALID_DECLARATION,
+					});
+					return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:978]:
```suggestion
					return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 22]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				this.nextToken();
+
+				while (this.matchToken("Whitespace")) {
+					this.eatToken("Whitespace");
+				}
+				if (!this.matchToken("Colon")) {
+					this.unexpectedDiagnostic({
+						description: descriptions.CSS_PARSER.INVALID_DECLARATION,
+					});
+					return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:994]:
```suggestion
					return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 23]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				}
+
+				const name = currentToken.value;
+				const start = this.getPosition();
+				let important = false;
+				let value: Array<AnyCSSValue | null> = [];
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:984]:
```suggestion
				let value: Array<AnyCSSValue | undefined> = [];
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 24]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+					return "RightCurlyBracket";
+				case "LeftParen":
+					return "RightParen";
+				case "LeftSquareBracket":
+					return "RightSquareBracket";
+				default: {
+					this.unexpectedDiagnostic({
+						description: descriptions.CSS_PARSER.INVALID_BLOCK_START,
+					});
+					return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:1058]:
```suggestion
					return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 25]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				}
+			}
+		}
+
+		getBlockEndTokenType(token: AnyCSSToken): keyof Tokens | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:1046]:
```suggestion
		getBlockEndTokenType(token: AnyCSSToken): keyof Tokens | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 26]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+					return "{";
+				case "LeftParen":
+					return "(";
+				case "LeftSquareBracket":
+					return "[";
+				default: {
+					this.unexpectedDiagnostic({
+						description: descriptions.CSS_PARSER.INVALID_BLOCK_START,
+					});
+					return null;
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:1041]:
```suggestion
					return undefined;
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 27]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+						);
+					}
+					const parsedValue = this.parseComponentValue();
+					parsedValue && value.push(parsedValue);
+				}
+			}
+			return null;
+		}
+
+		getBlockStartTokenValue(token: AnyCSSToken): string | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:1029]:
```suggestion
		getBlockStartTokenValue(token: AnyCSSToken): string | undefined {
```

File: packages/@romefrontend/css-parser/parse.ts [hunk 28]
```diff
@@ -589,5 +607,457 @@ export const createCSSParser = createParser((ParserCore) =>
...
+				start,
+				{
+					type: "CSSBlock",
+					startingTokenValue,
+					value,
+				},
+			);
+		}
+
+		parseComponentValue(): AnyCSSValue | null {
```
Comment (sebmck) [inline packages/@romefrontend/css-parser/parse.ts:818]:
```suggestion
		parseComponentValue(): AnyCSSValue | undefined {
```

File: scripts/vendor/rome.cjs [hunk 29]
```diff
@@ -19006,7 +19006,7 @@ const ___R$project$rome$$romefrontend$compiler$scope$globals_ts = {
 		"CSSRuleList",
 		"CSSStyleDeclaration",
 		"CSSStyleRule",
-		"CSSStyleSheet",
+		"CSSStylesheet",
```
Comment (sebmck) [inline scripts/vendor/rome.cjs:19009]: Seems to be a bad find+replace

```suggestion
		"CSSStyleSheet",
```

File: packages/@romefrontend/compiler/scope/globals.ts [hunk 30]
```diff
@@ -324,7 +324,7 @@ export const browser = [
 	"CSSRuleList",
 	"CSSStyleDeclaration",
 	"CSSStyleRule",
-	"CSSStyleSheet",
+	"CSSStylesheet",
```
Comment (sebmck) [inline packages/@romefrontend/compiler/scope/globals.ts:327]: Seems to be a bad find+replace

```suggestion
	"CSSStyleSheet",
```

**Pull request merged**